Parse tokens from all dependencies instead of single files in the language server

JoshuaBatty opened the pull request:
This PR now traverses the submodules in a project and collects those tokens into the `TokenMap`. 

I've moved a lot of the logic that was previously in `document.rs` into `session.rs`. Also ended moving / simplifying a bit of code so it's easier to read.

Finally, I've changed our `TokenMap` type to use a `DashMap` instead of a `HashMap`.

closes #2318 

partially addresses #324

mitchmindtree commented on sway-lsp/src/core/session.rs:
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
+        }
+    }
+
+    /// Check if the code editor's cursor is currently over an of our collected tokens
```
```suggestion
    /// Check if the code editor's cursor is currently over one of our collected tokens.
```

mitchmindtree commented on sway-lsp/src/core/session.rs:
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
+        // Look up the tokens TypeId
+        match utils::token::type_id(token) {
+            Some(type_id) => {
+                tracing::info!("type_id = {:#?}", type_id);
```
Is this left over from debugging?

mitchmindtree commented on sway-lsp/src/core/session.rs:
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
+
+                // Use the TypeId to look up the actual type
+                let type_info = sway_core::type_engine::look_up_type_id(type_id);
+                tracing::info!("type_info = {:#?}", type_info);
```
Same here

JoshuaBatty replied:
Yep good spots

mitchmindtree commented on sway-lsp/src/server.rs:
```diff
@@ -70,24 +72,27 @@ fn capabilities() -> ServerCapabilities {
         if self.config.parsed_tokens_as_warnings {
-            if let Some(document) = self.session.documents.get(uri.path()) {
-                let diagnostics = debug::generate_warnings_for_parsed_tokens(document.token_map());
+            //let diagnostics = debug::generate_warnings_for_parsed_tokens(token_map);
```
Left over from debugging?

JoshuaBatty replied:
I actually had the wrong line commented out, thanks for this.

mitchmindtree reviewed (commented):
Nice one, looking much cleaner! Just a couple nits, otherwise happy to approve :heavy_check_mark:

JoshuaBatty commented:
> Nice one, looking much cleaner! Just a couple nits, otherwise happy to approve ✔️

Thanks, @mitchmindtree just addressed those nits, should be good to go.

mitchmindtree reviewed (approved):
:shipit:
